### ISSUE TITLE
[$500] Desktop - Draft message offline isn't saved after app reload if user remains in same chat

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#33938]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.21-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Open Desktop App
2. Go to any chat
3. Go offline
4. Enter any message and do not send it
5. Close Desktop App without changing chat
6. Go online
7. Start Desktop app again

## Expected Result:
A draft message entered offline is saved in the chat composer field

## Actual Result:
A draft message entered offline is not saved after restarting the app and going online if the user remains in the same chat

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/16b4edfc-6e58-4b0e-8565-866a4bd6cbbd

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011a94f38191440005</li>
        <li>Upwork Job ID: 1742849765768896512</li>
        <li>Last Price Increase: 2024-01-04</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 28121676</li>
</ul></li>
    </ul>
</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Draft message offline isn't saved after app reload if user remains in same chat

### What is the root cause of that problem?
We are getting the draft comment manually here from onyx when the page is rendered here
https://github.com/Expensiy/App/blob/dc3e19bdcc95023d74cd7fdd433f6208707eaaf8/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L119-L125
So when the onyx data is not ready when rendering (when navigating directly to the chat on opening the app) it will be empty
### What changes do you think we should make in order to solve the problem?

We should subscribe to the data instead
```

        comment: {
            key: ({reportID}) => `${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}`,
            initialValue: '',
        },
```
### What alternative solutions did you explore? (Optional)

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Desktop - Draft message offline isn't saved after app reload if user remains in same chat
### What is the root cause of that problem?
- This error only appears when we open the app by the deep link https://staging.new.expensify.com/
- When we open the app by the above deeplink, the screen stacks is something like ```[{name: "home"}, {name: CentralPaneNavigator, state: {routes: {name: Report, params: {reportID: " "}}}]```
- As we can see, the first time, the app will render the report screen with reportID params is " ", and because we have the bellow logic to set the initial value for the ```value``` state
https://github.com/Expensify/App/blob/5fedab802a07d672efc544839f90ba65513f5d44/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L119-L125
so the initial value of the ```value``` state will be " " (because the reportId is " ", so getDraftComment(" ") will return " " as well)
- That leads to the current bug.
### What changes do you think we should make in order to solve the problem?
- We can make sure that the component [ComposerWithSuggestions](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js) is only rendered if the ```reportID``` is not empty.
- For example, do it by updating:
https://github.com/Expensify/App/blob/5fedab802a07d672efc544839f90ba65513f5d44/src/pages/home/report/ReportActionCompose/ReportActionCompose.js#L413
to:
```
                                    {isReportReadyForDisplay &&  <ComposerWithSuggestions ... />}
```

### What alternative solutions did you explore? (Optional)
- NA

**Comment by Christinadobrzyn** [comment]
I can reproduce this, I think it might be part of the Vip-vsb project so adding it there. I'm not sure if this needs to be internal but I think it can be external. 

@rushatgabhane feel free to continue with the proposal review!

**Comment by Christinadobrzyn** [comment]
I'm going to be ooo next week so adding a buddy to monitor this while I'm away. 

Status- we're reviewing proposals

**Comment by joekaufmanexpensify** [comment]
Sounds good!

**Comment by rushatgabhane** [comment]
@FitseTLT i think the root cause is not correct because `getDraftComment()` gets data from Onyx

**Comment by rushatgabhane** [comment]
@DylanDylann i think you have the correct root cause. 

` {isReportReadyForDisplay &&  <ComposerWithSuggestions ... />}`

Will waiting for `reportId` lead to a delayed composer? We want composer to display even if the report isn't ready yet.
So could we display empty composer initially and then fill it with draft comment once the report is ready? What do you think?

**Comment by DylanDylann** [comment]
@rushatgabhane I my proposal, I also suggested that we can use ```  {reportID !== '0' &&  <ComposerWithSuggestions ... />}``` so the composer is not delayed

**Comment by rushatgabhane** [comment]
Ohh okay that would work, thank you 👍
🎀 👀 🎀 https://github.com/Expensify/App/issues/33938#issuecomment-1877354537

**Comment by rushatgabhane** [comment]
Actually, @DylanDylann even if reportId is 0 can we show it immediately and put the draft text once it's available

**Comment by DylanDylann** [comment]
@rushatgabhane 

> can we show it immediately and put the draft text once it's available

How can we do that? Because once the ComposerWithSuggestions is rendered, the below will be called

https://github.com/Expensify/App/blob/5fedab802a07d672efc544839f90ba65513f5d44/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.js#L119-L125
so it will set the value to ``''```

**Comment by situchan** [comment]
Is this dupe of https://github.com/Expensify/App/issues/31705?

**Comment by DylanDylann** [comment]
@situchan No. It is not a dupe. The reproduction steps, the RCA and selected solution are different

**Comment by rushatgabhane** [comment]
@joekaufmanexpensify i crossed it because we need a more refined proposal 

https://github.com/Expensify/App/issues/33938#issuecomment-1884210465

**Comment by rushatgabhane** [comment]
> How can we do that? Because once the ComposerWithSuggestions is rendered, the below will be called

@DylanDylann can we put it under useEffect?

**Comment by rushatgabhane** [comment]
Can we re render once the draft comment is available?
```
useEffect(
...
[getDraftComment(reportID)])
```

**Comment by DylanDylann** [comment]
@rushatgabhane 

- I don`t think so. Because the RCA is that the report `s composer always has the initial value state is ```''```. So if we re-render, the bug is not fixed

**Comment by DylanDylann** [comment]
> Can we pass the correct reportId to the composer once after reportId is ready?

Yes. That is the same as my solution

**Comment by joekaufmanexpensify** [comment]
@rushatgabhane Are we close to being able to select a proposal here?

**Comment by joekaufmanexpensify** [comment]
Pending further input from @rushatgabhane

**Comment by rushatgabhane** [comment]
> Can we pass the correct reportId to the composer once after reportId is ready?
> Yes. This is the same as my solution

@DylanDylann no because the solution won't render the composer when reportId is `0`.

**Comment by rushatgabhane** [comment]
@puneetlath let's go with @DylanDylann's [proposal](https://github.com/Expensify/App/issues/33938#issuecomment-1877354537)
C+ reviewed 🎀 👀🎀

**Comment by joekaufmanexpensify** [comment]
We should be able to proceed with the proposal here soon.

**Comment by joekaufmanexpensify** [comment]
Reached out to Puneet to confirm if this proposal is good

**Comment by joekaufmanexpensify** [comment]
Still pending confirmation about proposal

**Comment by joekaufmanexpensify** [comment]
@puneetlath mind signing off on [this proposal](https://github.com/Expensify/App/issues/33938#issuecomment-1894878067) when you have a sec?

**Comment by puneetlath** [comment]
Sorry for the delay. Assigned.

**Comment by DylanDylann** [comment]
@rushatgabhane PR https://github.com/Expensify/App/pull/35213 is ready to review

**Comment by joekaufmanexpensify** [comment]
[Bumped in Slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1707144177153309)

**Comment by rushatgabhane** [comment]
Hey, looks like there was a dupe issue - https://github.com/Expensify/App/issues/34667
And it was fixed before us.

**Comment by DylanDylann** [comment]
Please note that the solution in this issue and the solution in #34667 are the same and the solution https://github.com/Expensify/App/issues/33938#issuecomment-1877354537 came first. And the [PR](https://github.com/Expensify/App/pull/35213) was delayed reviewing by a few weeks

**Comment by joekaufmanexpensify** [comment]
Interesting. So the PR we opened in this issue isn't needed at all anymore. Is that right?

**Comment by DylanDylann** [comment]
@joekaufmanexpensify Yes

**Comment by joekaufmanexpensify** [comment]
Got it. Thanks! Looking into this more.

**Comment by joekaufmanexpensify** [comment]
We discussed this internally, and here's where we landed: 

- @DylanDylann you'll get the full $500 bounty here since you were hired and raised the PR. This will be paid via Upwork. 
- @rushatgabhane, we're going with $0 for the C+ bounty here. Specifically because review on the PR was delayed by 10+ days here. If it hadn't been, this issue's PR would've likely been merged before we hired anyone in the other issue, meaning it would've been closed without compensation for the C+ there. Since we're paying the C+ bounty there, we're going to omit it here. 

LMK if you have any questions!
